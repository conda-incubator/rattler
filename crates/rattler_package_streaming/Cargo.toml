[package]
name = "rattler_package_streaming"
version = "0.20.4"
edition.workspace = true
authors = ["Bas Zalmstra <zalmstra.bas@gmail.com>"]
description = "Extract and stream of Conda package archives"
categories.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true
readme.workspace = true

[dependencies]
bzip2 = { workspace = true }
chrono = { workspace = true }
futures-util = { workspace = true }
num_cpus = { workspace = true }
rattler_conda_types = { path="../rattler_conda_types", version = "0.21.0", default-features = false }
rattler_digest = { path="../rattler_digest", version = "0.19.3", default-features = false }
rattler_networking = { path="../rattler_networking", version = "0.20.2", default-features = false }
reqwest = { workspace = true, features = ["stream"], optional = true }
reqwest-middleware = { workspace = true, optional = true }
serde_json = { workspace = true }
tar = { workspace = true }
tempfile = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["fs"] }
tokio-util = { workspace = true, features = ["io-util"] }
url = { workspace = true }
zip = { workspace = true, features = ["deflate", "time"] }
zstd = { workspace = true, features = ["zstdmt"] }

[features]
default = ["native-tls"]
native-tls = ["rattler_networking/native-tls"]
rustls-tls = ["rattler_networking/rustls-tls"]
wasm = ["zstd/wasm"]
reqwest = ["dep:reqwest-middleware", "dep:reqwest"]

[dev-dependencies]
tokio = { workspace = true, features = ["rt", "macros", "rt-multi-thread"] }
walkdir = { workspace = true }
rstest = { workspace = true }
rstest_reuse = { workspace = true }
assert_matches = { workspace = true }
