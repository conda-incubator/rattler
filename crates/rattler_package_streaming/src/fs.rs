//! Functions to extracting or stream a Conda package from a file on disk.

use crate::{read, seek};
use crate::{ExtractError, ExtractResult};
use rattler_conda_types::package::ArchiveType;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

/// Extracts the contents a `.tar.bz2` package archive at the specified path to a directory.
///
/// ```rust,no_run
/// # use std::path::Path;
/// use rattler_package_streaming::fs::extract_tar_bz2;
/// let _ = extract_tar_bz2(
///     Path::new("conda-forge/win-64/python-3.11.0-hcf16a7b_0_cpython.tar.bz2"),
///     Path::new("/tmp"))
///     .unwrap();
/// ```
pub fn extract_tar_bz2(archive: &Path, destination: &Path) -> Result<ExtractResult, ExtractError> {
    let file = File::open(archive)?;
    crate::read::extract_tar_bz2(file, destination)
}

/// Extracts the contents a `.conda` package archive at the specified path to a directory.
///
/// ```rust,no_run
/// # use std::path::Path;
/// use rattler_package_streaming::fs::extract_conda;
/// let _ = extract_conda(
///     Path::new("conda-forge/win-64/python-3.11.0-hcf16a7b_0_cpython.conda"),
///     Path::new("/tmp"))
///     .unwrap();
/// ```
pub fn extract_conda(archive: &Path, destination: &Path) -> Result<ExtractResult, ExtractError> {
    let file = File::open(archive)?;
    crate::read::extract_conda(file, destination)
}

/// Extracts the contents a package archive at the specified path to a directory. The type of
/// package is determined based on the file extension of the archive path.
///
/// ```rust,no_run
/// # use std::path::Path;
/// use rattler_package_streaming::fs::extract;
/// let _ = extract(
///     Path::new("conda-forge/win-64/python-3.11.0-hcf16a7b_0_cpython.conda"),
///     Path::new("/tmp"))
///     .unwrap();
/// ```
pub fn extract(archive: &Path, destination: &Path) -> Result<ExtractResult, ExtractError> {
    match ArchiveType::try_from(archive).ok_or(ExtractError::UnsupportedArchiveType)? {
        ArchiveType::TarBz2 => extract_tar_bz2(archive, destination),
        ArchiveType::Conda => extract_conda(archive, destination),
    }
}

/// Extracts a specified directory from .tar.bz2 into a destination folder.
///
/// ```rust,no_run
/// # use std::path::Path;
/// use rattler_package_streaming::fs::extract_directory_from_tar_bz2;
/// let _ = extract_directory_from_tar_bz2(Path::new("archive/location"), Path::new("directory_to_extract"), Path::new("destination/directory"))
/// ```
pub fn extract_directory_from_tar_bz2(
    archive_path: &Path,
    directory_to_extract: &Path,
    dest_directory: &Path,
) -> Result<(), std::io::Error> {
    let reader = std::fs::File::open(archive_path)?;
    let mut archive = read::stream_tar_bz2(reader);

    for entry in archive.entries()? {
        let mut entry = entry?;
        let path = entry.path()?;

        if let Ok(stripped_path) = path.strip_prefix(directory_to_extract) {
            let dest_file = dest_directory.join(stripped_path);
            if let Some(parent_folder) = dest_file.parent() {
                if !parent_folder.exists() {
                    std::fs::create_dir_all(parent_folder)?;
                }
            }
            let mut dest_file = std::fs::File::create(dest_file)?;
            std::io::copy(&mut entry, &mut dest_file)?;
        }
    }
    Ok(())
}

/// Extracts a specified directory from .conda archive into a destination folder.
///
/// ```rust,no_run
/// # use std::path::Path;
/// use rattler_package_streaming::fs::extract_directory_from_tar_bz2;
/// let _ = extract_directory_from_conda(Path::new("archive/location"), Path::new("directory_to_extract"), Path::new("destination/directory"))
/// ```
pub fn extract_directory_from_conda(
    archive_path: &Path,
    directory_to_extract: &Path,
    dest_directory: &Path,
) -> Result<(), std::io::Error> {
    let reader = std::fs::File::open(archive_path)?;
    let buf_reader = BufReader::new(reader);

    let mut archive = if directory_to_extract.starts_with("info") {
        seek::stream_conda_info(buf_reader).expect("Could not open conda file")
    } else {
        todo!("Not implemented yet");
    };

    for entry in archive.entries()? {
        let mut entry = entry?;
        let path = entry.path()?;
        if let Ok(stripped_path) = path.strip_prefix(directory_to_extract) {
            let dest_file = dest_directory.join(stripped_path);
            if let Some(parent_folder) = dest_file.parent() {
                if !parent_folder.exists() {
                    std::fs::create_dir_all(parent_folder)?;
                }
            }
            let mut dest_file = std::fs::File::create(dest_file)?;
            std::io::copy(&mut entry, &mut dest_file)?;
        }
    }
    Ok(())
}
