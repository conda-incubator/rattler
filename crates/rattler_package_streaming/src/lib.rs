//! This crate provides the ability to extract a package archive or specific parts of it.

use std::path::Path;

pub mod read;
pub mod seek;

#[cfg(feature = "reqwest")]
pub mod reqwest;

#[cfg(feature = "tokio")]
pub mod tokio;

/// An error that can occur when extracting a package archive.
#[derive(thiserror::Error, Debug)]
pub enum ExtractError {
    #[error("an io error occurred")]
    IoError(#[from] std::io::Error),

    #[error("could not create the destination path")]
    CouldNotCreateDestination(#[source] std::io::Error),

    #[error("invalid zip archive")]
    ZipError(#[from] zip::result::ZipError),

    #[error("a component is missing from the Conda archive")]
    MissingComponent,

    #[error("unsupported compression method")]
    UnsupportedCompressionMethod,

    #[cfg(feature = "reqwest")]
    #[error(transparent)]
    ReqwestError(::reqwest::Error),

    #[error("unsupported package archive format")]
    UnsupportedArchiveType,

    #[error("the task was cancelled")]
    Cancelled,
}

/// Describes the type of package archive. This can be derived from the file extension of a package.
#[derive(Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub enum ArchiveType {
    /// A file with the `.tar.bz2` extension.
    TarBz2,

    /// A file with the `.conda` extension.
    Conda,
}

impl ArchiveType {
    /// Tries to determine the type of a Conda archive from its filename.
    pub fn try_from(path: &Path) -> Option<ArchiveType> {
        let file_name = path.file_name()?.to_string_lossy();
        if file_name.ends_with(".conda") {
            Some(ArchiveType::Conda)
        } else if file_name.ends_with(".tar.bz2") {
            Some(ArchiveType::TarBz2)
        } else {
            None
        }
    }
}
