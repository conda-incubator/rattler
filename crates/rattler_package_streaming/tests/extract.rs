use rattler_package_streaming::{
    read::{extract_conda, extract_tar_bz2, parallel_extract_tar_bz2},
    ArchiveType,
};
use std::fs::File;
use std::path::{Path, PathBuf};

fn find_all_archives() -> impl Iterator<Item = PathBuf> {
    std::fs::read_dir(Path::new(env!("CARGO_MANIFEST_DIR")).join("data"))
        .unwrap()
        .filter_map(Result::ok)
        .map(|d| d.path())
}

#[test]
fn test_extract_conda() {
    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));
    println!("Target dir: {}", temp_dir.display());

    for file_path in
        find_all_archives().filter(|path| ArchiveType::try_from(path) == Some(ArchiveType::Conda))
    {
        println!("Name: {}", file_path.display());

        let target_dir = temp_dir.join(file_path.file_stem().unwrap());
        extract_conda(File::open(&file_path).unwrap(), &target_dir).unwrap();
    }
}

#[test]
fn test_stream_info() {
    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));
    println!("Target dir: {}", temp_dir.display());

    for file_path in
        find_all_archives().filter(|path| ArchiveType::try_from(path) == Some(ArchiveType::Conda))
    {
        println!("Name: {}", file_path.display());

        let mut info_stream =
            rattler_package_streaming::seek::stream_conda_info(File::open(&file_path).unwrap())
                .unwrap();

        let target_dir = temp_dir.join(format!(
            "{}-info",
            file_path.file_stem().unwrap().to_string_lossy().as_ref()
        ));

        info_stream.unpack(target_dir).unwrap();
    }
}

#[test]
fn test_extract_tar_bz2() {
    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));
    println!("Target dir: {}", temp_dir.display());

    for file_path in
        find_all_archives().filter(|path| ArchiveType::try_from(path) == Some(ArchiveType::TarBz2))
    {
        println!("Name: {}", file_path.display());

        let target_dir = temp_dir.join(file_path.file_stem().unwrap());
        extract_tar_bz2(File::open(&file_path).unwrap(), &target_dir).unwrap();
    }
}

#[test]
fn test_extract_tar_parallel() {
    let temp_dir = Path::new(env!("CARGO_TARGET_TMPDIR"));
    println!("Target dir: {}", temp_dir.display());

    for file_path in
        find_all_archives().filter(|path| ArchiveType::try_from(path) == Some(ArchiveType::TarBz2))
    {
        println!("Name: {}", file_path.display());

        let target_dir = temp_dir.join(format!(
            "{}-parallel",
            file_path.file_stem().unwrap().to_string_lossy().as_ref()
        ));
        parallel_extract_tar_bz2(
            File::open(&file_path).unwrap(),
            &target_dir,
            1024 * 1024 * 10,
        )
        .unwrap();
    }
}
