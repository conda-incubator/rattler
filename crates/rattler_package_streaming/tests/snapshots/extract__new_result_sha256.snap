---
source: crates/rattler_package_streaming/tests/extract.rs
expression: "&format!(\"{:x}\", new_result.sha256)"
---
6a5d6d8a1a7552dbf8c617312ef951a77d2dac09f2aeaba661deebce603a7a97
